src/policy/marksweepspace/native_ms/block.rs

```diff
@@ -0,0 +1,450 @@
+            }
+            BlockState::Marked => {
+                // The block is live.
+                false
+            }
+        }
+    }
+
+    /// Sweep the block. This is done either lazily in the allocation phase, or eagerly at the end of a GC.
+    pub fn sweep<VM: VMBinding>(&self) {
+        // The important point here is that we need to distinguish cell address, allocation address, and object reference.
+        // We only know cell addresses here. We do not know the allocation address, and we also do not know the object reference.
+        // The mark bit is set for object references, and we need to use the mark bit to decide whether a cell is live or not.
+
+        // We haven't implemented for malloc/free cases, for which we do not have mark bit. We could use valid object bit instead.
+        if cfg!(feature = "malloc_native_mimalloc") {
+            unimplemented!()
+        }
+
+        // Check if we can treat it as the simple case: cell address === object reference.
+        // If the binding does not use allocation offset, and they use the same allocation alignment which the cell size is aligned to,
+        // then we have cell address === allocation address.
+        // Furthermore, if the binding does not have an offset between allocation and object reference, then allocation address === cell address.
+        if !VM::USE_ALLOCATION_OFFSET
+            && VM::MAX_ALIGNMENT == VM::MIN_ALIGNMENT
+            && crate::util::conversions::raw_is_aligned(
+                self.load_block_cell_size(),
+                VM::MAX_ALIGNMENT,
+            )
+            && VM::VMObjectModel::UNIFIED_OBJECT_REFERENCE_ADDRESS
+        {
+            // In this case, we can use the simplest and the most efficicent sweep.
+            self.simple_sweep::<VM>()
+        } else {
+            // Otherwise we fallback to a generic but slow sweep. This roughly has ~10% mutator overhead for lazy sweeping.
+            self.naive_brute_force_sweep::<VM>()
+        }
+    }
+
+    /// This implementation uses object reference and cell address interchangably. This is not correct for most cases.
+    /// However, in certain cases, such as OpenJDK, this is correct, and efficient. See the sweep method for the invariants
+    /// that we need to use this method correctly.
+    fn simple_sweep<VM: VMBinding>(&self) {
+        let cell_size = self.load_block_cell_size();
+        let mut cell = self.start();
+        let mut last = unsafe { Address::zero() };
+        while cell + cell_size <= self.start() + Block::BYTES {
+            // We know the cell and the object reference is the same.
+            if !VM::VMObjectModel::LOCAL_MARK_BIT_SPEC
+                .is_marked::<VM>(ObjectReference::from_raw_address(cell), Ordering::SeqCst)
+            {
+                // clear alloc bit if it is ever set.
+                #[cfg(feature = "global_alloc_bit")]
+                crate::util::alloc_bit::ALLOC_SIDE_METADATA_SPEC.store_atomic::<u8>(
```

Comment:
It is preferable to call `crate::util::alloc_bit::bzero_alloc_bit` to ensure all accesses to the VO-bit metadata are encapsulated.

Reply:
Done